### DISCUSSION TITLE
Can you move an `App` into an `HttpServer::new`?

### DISCUSSION THREAD

**Discussion by alshdavid** [actix/actix-web#3417] [Q&A] [answered]
Not a major issue, just a code style preference brought over from http frameworks I have used in other languages. I am wondering why I cannot move my `App` into `HttpServer::new`.

What I'd like to do is define my app like this:

```rust
#[tokio::main]
async fn main()  -> std::io::Result<()> {
  let mut app = App::new();
  
  app.route("/", web::get().to(manual_hello));
  app.route("/foo", web::get().to(handler_foo));
  app.route("/bar", web::get().to(handler_bar));

  HttpServer::new(move || app)
    .bind(("127.0.0.1", 8080))?.run().await
}

async fn manual_hello() -> impl Responder {
  HttpResponse::Ok().body("Hey there!")
}
```

However it appears that `the trait `Clone` is not implemented for `App<actix_web::app_service::AppEntry>`. Which I understand but I wonder if there is a way to use Actix with the pattern described above?

**Comment by robjtede** [comment, marked as answer]
The closure passed to `HttpServer::new` is an app **factory** and `App` doesn't impl Clone so it's not possible to just clone the app into the re-usable closure.

You can however [return an app from a function](https://github.com/actix/actix-web/wiki/FAQ#how-can-i-return-app-from-a-function--why-is-appentry-private) or use [`App::configure()`](https://docs.rs/actix-web/4/actix_web/struct.App.html#method.configure) to split up or share routing setups.
